### PR TITLE
Added support for unwrapped Azure API error responses

### PR DESCRIPTION
**Pull request by michalpristas**
Should solve problem described here: https://github.com/Azure/azure-sdk-for-go/issues/748
The problem is that result of API call is ServiceError itself, not ServiceError wrapped in error object. Which causes autorest to return 'Unknown' error instead.

Fixes: 748

### CODE REVIEW

File: autorest/azure/azure.go [hunk 1]
```diff
@@ -165,7 +165,13 @@ func WithErrorUnlessStatusCode(codes ...int) autorest.RespondDecorator {
...
 					return fmt.Errorf("autorest/azure: error response cannot be parsed: %q error: %v", b.String(), decodeErr)
 				} else if e.ServiceError == nil {
-					e.ServiceError = &ServiceError{Code: "Unknown", Message: "Unknown service error"}
+					// Check if error is unwrapped ServiceError
+					var se *ServiceError
+					if err := json.Unmarshal(b.Bytes(), &se); err == nil && se.Message != "" {
+						e.ServiceError = se
+					} else {
+						e.ServiceError = &ServiceError{Code: "Unknown", Message: "Unknown service error"}
+					}
```
Comment (marstr) [inline autorest/azure/azure.go]: I believe this should be updated to:

``` Go
var se ServiceError
if err := json.Unmarshal(b.Bytes(), &se); err == nil && se.Message != "" {
    e.ServiceError = &se
} else {
    e.ServiceError = &ServiceError{
        Code: "Unknown",
        Message: "Unknown service error",
    }
}
```

File: autorest/azure/azure_test.go [hunk 2]
```diff
@@ -333,6 +333,64 @@ func TestWithErrorUnlessStatusCode_NoAzureError(t *testing.T) {
...
+	mocks.SetResponseHeader(r, HeaderRequestID, uuid)
+	r.Request = mocks.NewRequest()
+	r.StatusCode = http.StatusInternalServerError
+	r.Status = http.StatusText(r.StatusCode)
+
+	err := autorest.Respond(r,
+		WithErrorUnlessStatusCode(http.StatusOK),
+		autorest.ByClosing())
+	if err == nil {
+		t.Fatalf("azure: returned nil error for proper error response")
```
Comment (marstr) [inline autorest/azure/azure_test.go]: Using `t.Fatalf` ends all test execution. Please update these calls to be of the pattern:

``` Go
t.Logf("[your stuff]", [any args you need])
t.Fail()
```

File: autorest/azure/azure_test.go [hunk 3]
```diff
@@ -333,6 +333,64 @@ func TestWithErrorUnlessStatusCode_NoAzureError(t *testing.T) {
...
+	if expected := "Azure is having trouble right now."; azErr.ServiceError.Message != expected {
+		t.Fatalf("azure: got wrong Message=%s Expected=%s", azErr.Message, expected)
+	}
+
+	if expected := uuid; azErr.RequestID != expected {
+		t.Fatalf("azure: wrong request ID in error. expected=%q; got=%q", expected, azErr.RequestID)
+	}
+
+	details, _ := json.Marshal(*azErr.ServiceError.Details)
+	if string(details) != `[{"code":"conflict1","message":"error message1"},{"code":"conflict2","message":"error message2"}]` {
```
Comment (marstr) [inline autorest/azure/azure_test.go]: A Go idiom is to print out "got" and "want" in this case, one would expect the Unit Test Log to read the following on failure:

```
got:  "{contents of details}"
want: "[{"code":"conflict1","message":"error message1"},{"code":"conflict2","message":"error message2"}]"
```

Mind refactoring this?

Comment (michalpristas) [reply autorest/azure/azure_test.go]: i will prepare fix today

**Comment by michalpristas** [comment]
@marstr is something missing?

**Comment by marstr** [comment]
Sorry, to leave you hanging @michalpristas! Thanks for making the recommended changes. Merging this in now.

**Pull request merged**